File: 0127-word-ladder/0127-word-ladder.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        HashSet<String> s = new HashSet<>();
        for(String i:wordList)
            s.add(i);
        Queue<String> q = new LinkedList<>();
        q.add(beginWord);
        int cnt = 0;
        while(!q.isEmpty()){
            cnt++;
            int n = q.size();
            while(n != 0){
                String u = q.remove();
                for(int i = 0; i < u.length(); i++)
                    for(char j = 'a'; j <= 'z'; j++){
                        String v = u.substring(0, i) + j + u.substring(i + 1);
                        // System.out.println(v);
                        if(s.contains(v)){
                            s.remove(v);
                            q.add(v);
                            if(endWord.equals(v))
                                return cnt + 1;
                        }
                    }
                n--;
            }
        }
        return 0;
    }
}

File: 0127-word-ladder/0127-word-ladder.cpp
class Solution {
public:
    int ladderLength(string beginWord, string endWord, vector<string>& wordList) {
        unordered_set<string> st(wordList.begin(), wordList.end());
        queue<pair<string, int>> q;
        q.push({beginWord, 1});
        st.erase(beginWord);

        while(!q.empty()){
            string word = q.front().first;
            int lvl = q.front().second;
            q.pop();

            for(int i = 0; i < word.length(); i++){
                for(char ch = 'a'; ch <= 'z'; ch++){
                    string temp = word;
                    temp[i] = ch;
                    if(st.count(temp)){
                        if(temp == endWord){
                            return lvl + 1;
                        }
                        q.push({temp, lvl + 1});
                        st.erase(temp);
                    }
                }
            }
        }

        return 0;
    }
};